File: leetcode-everyday/搜索算法/47-全排列II/47-全排列II.cpp
class Solution {
private:
    void dfs(vector<int>& nums, vector<vector<int>>& ans, vector<int>& path, vector<bool>& visited) {
        if(path.size() == nums.size()) {
            ans.push_back(path);
            return;
        }
        for(int i = 0; i < nums.size(); ++i) {
            if(visited[i]) continue;
            if(i >= 1 && nums[i] == nums[i-1] && !visited[i-1]) continue;
            visited[i] = true;
            path.push_back(nums[i]);
            dfs(nums, ans, path, visited);
            path.pop_back();
            visited[i] = false;
        }
    }
public:
    vector<vector<int>> permuteUnique(vector<int>& nums) {
        int m = nums.size();
        vector<bool> visited(m, false);
        vector<vector<int>> ans;
        vector<int> path;
        // sort(nums.begin(), nums.end(), [](int a, int b) {
        //     return a < b;
        // });
        sort(nums.begin(), nums.end());
        dfs(nums, ans, path, visited);
        return ans;
    }
};

File: leetcode-everyday/搜索算法/47-全排列II/47-全排列II.js
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
 var permuteUnique = function(nums) {
  let m = nums.length;
  const ans = [];
  const visited = new Array(m).fill(false);
  nums.sort((a,b) => a-b);
  dfs(nums, ans, [], visited);
  return ans;
};

function dfs(nums, ans, path, visited) {
  if(path.length === nums.length) {
      ans.push([...path]);
      return;
  }
  for(let i = 0; i < nums.length; ++i) {
      if(visited[i] || (i >= 1 && nums[i] === nums[i-1] && !visited[i-1])) {
          continue;
      }
      visited[i] = true;
      dfs(nums, ans, [...path, nums[i]], visited);
      visited[i] = false;
  }
}